### PR TITLE
Versions of algorithms implementing tighter LB update and other improvements

### PR DESCRIPTION
**Pull request by petrrysavy**

### CODE REVIEW

**Comment by ghamerly** [comment]
@petrrysavy can you provide a class hierarchy for all the algorithms (including the original set)?

File: fast_kmeans/modified_update_triangle_based_kmeans.h [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+		free();
+	}
+
+	virtual void initialize(Dataset const *aX, unsigned short aK, unsigned short *initialAssignment, int aNumThreads);
+	virtual void free();
+
+	/* This function claculated inner product of two vectors of dimension d.
+	 * The number of inner products is counted in the superclass.
+	 */
+	virtual inline double inner_product(double *a, double const *b) {
```
Comment (ghamerly) [inline fast_kmeans/modified_update_triangle_based_kmeans.h]: Why is `b` const and `a` is not? And why is this function not `const`?

**Comment by petrrysavy** [comment]
The class hierarchy:
class hierarchy

**Comment by ghamerly** [comment]
I have ported this code over to https://github.com/ghamerly/fast-kmeans/pull/1 and am closing this PR.

**Pull request closed without merging**